File: src/test/scala/javtang/UncheckedTest.scala
package javtang


import java.io.{FileReader, BufferedReader}
import java.util.concurrent.Callable

import org.scalatest.FunSuite
import com.google.common.base.Optional

class UncheckedTest extends FunSuite {
  test("Unchecked Test") {
    val fileContent: Optional[String] = Unchecked.call(new Callable[String] {
      def call: String = {
        return getFileContent("/home/itang/.bashrc")
      }
    })

    println(fileContent.get())

    try {
      Unchecked.call(new Callable[String] {
        def call: String = {
          return getFileContent("/home/itang/.bashrc_bad")
        }
      })
    }
    catch {
      case e: RuntimeException => {
        System.out.println(e.getMessage)
      }
    }
  }

  implicit def optionalToOption[String](optional: Optional[String]): Option[String] = {
    if (optional.isPresent) Some(optional.get()) else None
  }

  test("Unchecked for scala") {
    val result = Unchecked.call(new Callable[String] {
      def call() = "hello"
    })

    //FIXME 好吧, 在match pattern下implicit不起作用?
    optionalToOption(result) match {
      case Some(msg) => assert(msg === "hello")
      case None =>
    }
  }

  private def getFileContent(path: String): String = {
    val reader: BufferedReader = new BufferedReader(new FileReader(path))
    var line: String = null
    val sb: StringBuilder = new StringBuilder
    val lineSp: String = System.getProperty("line.separator")
    while (({
      line = reader.readLine;
      line
    }) != null) {
      sb.append(line).append(lineSp)
    }
    return sb.toString
  }
}

File: src/test/java/javtang/UncheckedTest.java
package javtang;

import static org.junit.Assert.assertTrue;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.concurrent.Callable;

import org.junit.Test;

public class UncheckedTest {

    @Test
    public void test() {
        String fileContent = Unchecked.call(new Callable<String>() {
            @Override
            public String call() throws Exception {
                return getFileContent("/home/itang/.bashrc");
            }
        });
        assertTrue(fileContent != null);

        try {
            Unchecked.call(new Callable<String>() {
                @Override
                public String call() throws Exception {
                    return getFileContent("/home/itang/.bashrc_bad");
                }
            });
        } catch (RuntimeException e) {
            System.out.println(e.getMessage());
        }
    }

    private static String getFileContent(String path) throws IOException {
        if ("/home/itang/.bashrc".equals(path)) {
            return "mock content";
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
            String line = null;
            StringBuilder sb = new StringBuilder();
            String lineSp = System.getProperty("line.separator");
            while ((line = reader.readLine()) != null) {
                sb.append(line).append(lineSp);
            }
            return sb.toString();
        }
    }
}
